hudi-client/hudi-java-client/src/main/java/org/apache/hudi/table/action/commit/JavaWriteHelper.java

```diff
@@ -69,17 +70,26 @@ public List<HoodieRecord<T>> deduplicateRecords(
     }).collect(Collectors.groupingBy(Pair::getLeft));
 
     final Schema schema = new Schema.Parser().parse(schemaStr);
-    return keyedRecords.values().stream().map(x -> x.stream().map(Pair::getRight).reduce((rec1, rec2) -> {
-      HoodieRecord<T> reducedRecord;
-      try {
-        reducedRecord =  merger.merge(rec1, schema, rec2, schema, props).get().getLeft();
-      } catch (IOException e) {
-        throw new HoodieException(String.format("Error to merge two records, %s, %s", rec1, rec2), e);
+    boolean sortBeforePrecombine = merger.useSortedMerge(props);
+    return keyedRecords.values().stream().map(hoodieRecords -> {
+      Stream<HoodieRecord<T>> recordStream;
+      if (sortBeforePrecombine) {
+        recordStream = hoodieRecords.stream().map(Pair::getRight).sorted(new SortedPrecombineComparator(schema, props));
```

Comment:
same question on the sorting here.

Reply:
Same as above. I've moved the work to a new PR: #9979